src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -21,133 +21,505 @@
+
+        if (sample == null)
+        {
+            AnnotatedMeasurement maybeNewSample = new AnnotatedMeasurement(measurementImpl(latencyMicros));
+            sample = samples.putIfAbsent(address, maybeNewSample);
+            if (sample == null)
+                sample = maybeNewSample;
+        }
+
+        if (measurementType == LatencyMeasurementType.READ && sample.millisSinceLastMeasure.get() > 0)
+            sample.millisSinceLastMeasure.lazySet(0);
+
+        sample.measurement.sample(latencyMicros);
+    }
+
+    @VisibleForTesting
+    protected void reset()
+    {
+        scores.clear();
+        samples.clear();
+    }
+
+    @VisibleForTesting
+    protected void updateScores()
     {
-        if (dynamicUpdateInterval != DatabaseDescriptor.getDynamicUpdateInterval())
+        if (!StorageService.instance.isGossipActive())
+            return;
+
+        if (!registered)
         {
-            dynamicUpdateInterval = DatabaseDescriptor.getDynamicUpdateInterval();
-            if (DatabaseDescriptor.isDaemonInitialized())
+            if (MessagingService.instance() != null)
             {
-                updateSchedular.cancel(false);
-                updateSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(update, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
+                MessagingService.instance().register(this);
+                registered = true;
             }
         }
 
-        if (dynamicResetInterval != DatabaseDescriptor.getDynamicResetInterval())
+        this.scores = calculateScores(samples);
+    }
+
+    /**
+     * This is generally expensive and is called periodically (semi-frequently) not on the fast path.
+     * The main concern here is generating garbage from the measurements (e.g. histograms in particular)
+     */
+    @VisibleForTesting
+    protected static Map<InetAddressAndPort, Double> calculateScores(Map<InetAddressAndPort, AnnotatedMeasurement> samples)
+    {
+        double maxLatency = 1;
+        HashMap<InetAddressAndPort, Double> newScores = new HashMap<>(samples.size());
+
+        // We're going to weight the latency for each host against the worst one we see, to
+        // arrive at sort of a 'badness percentage' for them. First, find the worst latency for each:
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry : samples.entrySet())
         {
-            dynamicResetInterval = DatabaseDescriptor.getDynamicResetInterval();
-            if (DatabaseDescriptor.isDaemonInitialized())
+            AnnotatedMeasurement annotatedMeasurement = entry.getValue();
+
+            // only compute the measurement, which probably generates the most garbage (e.g. for this Histogram),
+            // for endpoints that have been recently updated (millisSinceLastRequest) or somewhat recently requested
+            // for ranking.
+            if (annotatedMeasurement.millisSinceLastMeasure.get() < MIN_PROBE_INTERVAL_MS ||
```

Comment:
The problem with just checking `millisSinceLastRequest` here is that we don't calculate any scores at all until something is requested. I can fix this a few ways though, looking into it.

Reply:
Ok, I just ended up doing what you asked and I made the `millisSinceLastRequest` default to zero instead of maxint (so that way we'll calculate scores for all nodes for the first 10 minutes (like we did in the status quo), and then after that we stop calculating.